daemons/controld/controld_callbacks.c

```diff
@@ -118,6 +118,21 @@ peer_update_callback(enum crm_status_type type, crm_node_t * node, const void *d
         return;
     }
 
+    if ((type != crm_status_processes) && !is_remote) {
+        /*
+         * This is a hack until we can send to a nodeid and/or we fix node name lookups
+         * These messages are ignored in crmd_ha_msg_filter()
+         */
+        xmlNode *query = create_xml_node(NULL, __FUNCTION__);
+        crm_xml_add(query, F_TYPE, T_CRM);
+        crm_xml_add(query, T_CRM_OPERATION, "poke");
```

Comment:
The equivalent for the controller is create_request(), which will set F_CRM_TASK (the controller equivalent of F_STONITH_OPERATION, no T_CRM_OPERATION needed). You can define a PCMK__CONTROLD_CMD_POKE constant for "poke" in include/crm_internal.h.

Reply:
Actually, yes it was corosync-2.4.5, it should not be in the master. The problem occurs in both multi- amd unicast on a 3 nodes cluster. It's always appeared for me. TI only had to restart a cluster and call crm_node -l.

Reply:
It's also reproducable on older corosync versions (before the https://github.com/ClusterLabs/pacemaker/commit/8461509158e06365122dc741c527c83c94e966ce) by `crm_node -N node-id`. It would fail to show the `uname` for one of the nodes.

Reply:
> The problem occurs in both multi- and unicast on a 3-nodes cluster. It always appeared for me. I only had to restart a cluster on one of the nodes and call `crm_node -l`.

Is `nodelist` configured in corosync.conf with `ringX_addr` specified with IP addresses?

Reply:
Yes, if it's unicast, there is a nodelist with ringX_addr.

Reply:
And are ringX_addr options specified with  IP addresses rather than node names?

Reply:
Yes. It's for example
```
nodelist {
        node {
                ring0_addr: 10.162.32.211
                nodeid: 1
        }

        node {
                ring0_addr: 10.162.31.82
                nodeid: 2
        }

        node {
                ring0_addr: 10.162.32.182
                nodeid: 3
        }

}
```

Reply:
Thinking about the problem, I don't see how the fencer's approach is a good idea.

First, broadcasting our name for anything *other* than a process change seems like the exact wrong time. The other node can't receive the broadcast until it joins the process group. I'd expect the time to send a message would be when a crm_status_processes notification is received, and only when appeared is true, and the notification is for a different node.

Second, I don't see a point in broadcasting. I would think only the node that joined might need the poke. Even if a bunch of nodes are joining, sending a broadcast message for each one is adding overhead rather than saving any.

I do wonder whether there's a better way, since every time a node joins CPG, it will receive a poke from every other node, which seems like a good bit of overhead at join time (which is often a busy time already). I'd have to think more about how though.

Reply:
> Thinking about the problem, I don't see how the fencer's approach is a good idea.
> 
> First, broadcasting our name for anything _other_ than a process change seems like the exact wrong time. The other node can't receive the broadcast until it joins the process group. 

It appears that for most daemons,  crm_update_peer_proc() updates the state of the node in the membership cache and eventually calls peer_status_callback()  with crm_status_nstate. So it seems the original idea for fencer was basically to poke when the state of the node in the membership cache is actually being updated.

Conditions seem different here with controld though ...

> I'd expect the time to send a message would be when a crm_status_processes notification is received, and only when appeared is true, and the notification is for a different node.
> 
> Second, I don't see a point in broadcasting. I would think only the node that joined might need the poke. Even if a bunch of nodes are joining, sending a broadcast message for each one is adding overhead rather than saving any.

Makes sense to limit the conditions and scope.

> 
> I do wonder whether there's a better way, since every time a node joins CPG, it will receive a poke from every other node, which seems like a good bit of overhead at join time (which is often a busy time already). I'd have to think more about how though.

Otherwise probably there's anything actually useful for a node to tell/do when joining? :-)

Reply:
> > Thinking about the problem, I don't see how the fencer's approach is a good idea.
> > First, broadcasting our name for anything _other_ than a process change seems like the exact wrong time. The other node can't receive the broadcast until it joins the process group.
> 
> It appears that for most daemons, crm_update_peer_proc() updates the state of the node in the membership cache and eventually calls peer_status_callback() with crm_status_nstate. So it seems the original idea for fencer was basically to poke when the state of the node in the membership cache is actually being updated.
> 
> Conditions seem different here with controld though ...

Correct, tthe controller is different from all other daemons that use the peer cache in that regard -- the other daemons use "autoreap" so that changes in CPG result in a state change, and lost nodes are automatically removed from the cache. The controller disables autoreap so it can remove entries itself after e.g. confirming fencing.

So that makes some sense now, the fencer would see it differently. Unless I'm missing something, I do think it makes more sense for the controller to poke when a CPG join occurs. (I suspect the fencer could do the same, and it would be a little more efficient.)

> > I'd expect the time to send a message would be when a crm_status_processes notification is received, and only when appeared is true, and the notification is for a different node.
> > Second, I don't see a point in broadcasting. I would think only the node that joined might need the poke. Even if a bunch of nodes are joining, sending a broadcast message for each one is adding overhead rather than saving any.
> 
> Makes sense to limit the conditions and scope.
> 
> > I do wonder whether there's a better way, since every time a node joins CPG, it will receive a poke from every other node, which seems like a good bit of overhead at join time (which is often a busy time already). I'd have to think more about how though.
> 
> Otherwise probably there's anything actually useful for a node to tell/do when joining? :-)

I was thinking maybe the DC could send a list of known node names/IDs to newly joined nodes, and the node could update its peer cache with any missing names. But I'd have to think through the interactions with the various steps of the join process and DC election process, and make sure there's no chance of propagating outdated information.

Reply:
> So that makes some sense now, the fencer would see it differently. Unless I'm missing something, I do think it makes more sense for the controller to poke when a CPG join occurs. (I suspect the fencer could do the same, and it would be a little more efficient.)

Agreed.

> > > I do wonder whether there's a better way, since every time a node joins CPG, it will receive a poke from every other node, which seems like a good bit of overhead at join time (which is often a busy time already). I'd have to think more about how though.
> > 
> > 
> > Otherwise probably there's anything actually useful for a node to tell/do when joining? :-)
> 
> I was thinking maybe the DC could send a list of known node names/IDs to newly joined nodes, and the node could update its peer cache with any missing names. But I'd have to think through the interactions with the various steps of the join process and DC election process, and make sure there's no chance of propagating outdated information.

Yes, that should do but indeed needs more thinking...

I was kind of inclined to the easier idea of making the non-DC say anything at all to the newly joined node, given that DC already proactively talks to it :-)

Reply:
IMHO, the easier idea, where every alive node sends "Hello" (the poke message) to the newly joined should be error-prone and more resilient, rather than when it's done only by the DC.

Reply:
OK, so I think we're converging on a solution (at least for now): Non-DC nodes will send a poke directly to a node that joins controller CPG (crm_status_processes + appeared)

I am a little concerned about the overhead, since it's #nodes^2 messages at cluster start-up (to each joining node, it looks like all the other nodes joined CPG, so it will poke all of them too). As we shoot for scalability to hundreds of nodes that could become significant

Reply:
I would put it as a question of high availability vs high performance. That might indeed be a problem in the case of hundreds of nodes, but for our immediate purposes, the solution should be just good enough.